src/twisted/words/im/basesupport.py

```diff
@@ -121,13 +121,17 @@ def __init__(self, account, chatui, logonDeferred):
 
 
     def connectionMade(self):
-        self._protoBase.connectionMade(self)
+        _protoBase = getattr(self, '_protoBase', None)
```

Comment:
I don't know what to day.
Instead of this hack to work around mypy reporting, I think that is better to define `_protoBase` in AbstractClientMixin

Maybe
```python
class AbstractClientMixin:
    _protoBase = None
```

or 

```python
class AbstractClientMixin:
    _protoBase : Type[Protocol]
```

Reply:
The latter (`_protoBase : Type[Protocol]`) is by far the cleanest solution. The problem with working around it on access (`getattr()` etc.) is that every single access needs a workaround, which includes code outside of Twisted itself.

Also the extracted documentation will be incomplete (if the attribute is not declared) or wrong (if the attribute is declared with the wrong type, including `None` as an inferred type). It is possible to override this using `@type` in the docstring, but if this can be avoided the code will be easier to read and maintain.

To make it work in Python 3.5 you could cast `None` to `Type[Protocol]` or add a `type: ignore` comment. But maybe type annotations are a compelling enough reason to switch to Python 3.6 now rather than later.

Reply:
@adiroiban This won't work on Python 3.5:

```
class AbstractClientMixin:
    _protoBase : Type[Protocol]
```

Reply:
OK, I changed to this:

```
class AbstractClientMixin:
    _protoBase = None  # type: Type[Protocol]
```